### PR TITLE
WIP: Rework auth handling

### PR DESCRIPTION
**Pull request by dschmidt**
## Description
- replace abandoned oidc-client(-js) with actively maintained `oidc-client-ts` library.
- simplify auth handling

## Related Issue

- (hopefully :joy:) fixes https://github.com/owncloud/web/issues/7030

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] oidc-client-ts does not support implicit oauth grant (`token`), it requires usage of pkce (`code`).
     That unfortunately requires us to set a clientSecret in the auth config, needs to be checked for acceptance tests

### CODE REVIEW

File: changelog/5.5.0_2022-06-20/bugfix-not-logged-out-if-oC10-backend [hunk 1]
```diff
@@ -3,5 +3,6 @@ Bugfix: Not logged out if backend is ownCloud 10
 We've fixed an issue, where the user won't be logged out if the backend is ownCloud 10
 
 https://github.com/owncloud/web/pull/6939
+https://github.com/owncloud/web/pull/7072
```
Comment (pascalwengerter) [inline changelog/5.5.0_2022-06-20/bugfix-not-logged-out-if-oC10-backend:6]: Not sure when this PR was last rebased but by now it needs/deserves its own changelog item :)

Comment (pascalwengerter) [reply changelog/5.5.0_2022-06-20/bugfix-not-logged-out-if-oC10-backend:6]: Aye, cool

**Comment by pascalwengerter** [comment]
@kulmann could you check if this PR also fixes https://github.com/owncloud/web/issues/1749?

File: packages/web-runtime/package.json [hunk 2]
```diff
@@ -23,7 +23,7 @@
     "lodash-es": "^4.17.21",
     "luxon": "^2.4.0",
     "marked": "^4.0.12",
-    "oidc-client": "1.11.5",
+    "oidc-client-ts": "^2.0.5",
```
Comment (pascalwengerter) [inline packages/web-runtime/package.json:26]: 😻

File: packages/web-runtime/src/index.ts [hunk 3]
```diff
@@ -60,12 +64,17 @@ export const renderSuccess = (): void => {
...
+      getters['runtime/auth/isUserContextReady'] ||
+      getters['runtime/auth/isPublicLinkContextReady'],
+    async (newValue, oldValue) => {
+      // FIXME: a silent token renewal via iframe currently does a full re-init-cycle?! `oldValue` is false and `newValue` true in that case.
+      // can be triggered by setting short token lifetime (1 minute) in ocis and not having `offline_access` in the requested scopes.
       if (!newValue || newValue === oldValue) {
         return
       }
-      applications.forEach((application) => application.userReady(instance))
+      console.log('******* running ready() hook for all applications')
```
Comment (pascalwengerter) [inline packages/web-runtime/src/index.ts]: Do we want to be logging this?

File: packages/web-runtime/src/services/auth/authService.ts [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        )
+        try {
+          await this.userManager.updateContext(user.access_token)
+        } catch (e) {
+          console.error(e)
+          await this.handleAuthError(this.router.currentRoute)
+        }
+      })
+      this.userManager.events.addUserSignedIn(() => {
+        console.log('user signed in')
```
Comment (pascalwengerter) [inline packages/web-runtime/src/services/auth/authService.ts:81]: Dev leftover or worth keeping?

**Review by pascalwengerter — COMMENTED** [review]
Couldn't review it all but loved what I saw and minor questions, plus a green CI make me feel positive about it 🥳

File: packages/web-pkg/src/services/index.ts [hunk 5]
```diff
@@ -1,2 +1,2 @@
-export * from './client'
+export * from './client/index'
```
Comment (dschmidt) [inline packages/web-pkg/src/services/index.ts:1]: Huh wat?

File: packages/web-runtime/src/router/setupAuthGuard.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import { extractPublicLinkToken, isPublicLinkContext, isUserContext } from './index'
+import Router, { Route } from 'vue-router'
+import Vue from 'vue'
+import { contextRouteNameKey, queryItemAsString } from 'web-pkg/src/composables'
+
+export const setupAuthGuard = (router: Router) => {
+  router.beforeEach(async (to, from, next) => {
+    if (from && to.path === from.path && !hasContextRouteNameChanged(to, from)) {
+      // note: expect for the context route, query changes can never trigger re-init of the auth context
```
Comment (dschmidt) [inline packages/web-runtime/src/router/setupAuthGuard.ts:9]: `except`

File: packages/web-runtime/src/services/auth/authService.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+   */
+  public async signInSilentCallback() {
+    await this.userManager.signinSilentCallback()
+  }
+
+  public async handleAuthError(route: Route) {
+    if (isPublicLinkContext(this.router, route)) {
+      const token = extractPublicLinkToken(route)
+      this.publicLinkManager.clear(token)
+      await this.store.dispatch('runtime/auth/clearPublicLinkContext')
```
Comment (dschmidt) [inline packages/web-runtime/src/services/auth/authService.ts:160]: Should this go to `publicLinkManager.clear` maybe?

File: packages/web-runtime/src/services/auth/userManager.ts [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+    this.clientService.owncloudSdk.init(options)
+  }
+
+  private async fetchUserInfo(accessToken): Promise<void> {
+    const login = await this.clientService.owncloudSdk.getCurrentUser()
+    await this.fetchCapabilities({ accessToken })
+
+    // FIXME: Can be removed as soon as the uuid is integrated in the OCS api
+    // see https://github.com/owncloud/ocis/issues/3271
```
Comment (dschmidt) [inline packages/web-runtime/src/services/auth/userManager.ts:165]: I think the comment moved to a weird place when we added the roles here. I think it originally referred to https://github.com/owncloud/web/pull/7072/files#diff-7c13ed3f4e1b32127b202bef84f63748484b2cd21a50379a369a7c179137d149R206

We should probably elaborate a bit further here now...

File: packages/web-app-files/src/components/AppBar/CreateSpace.vue [hunk 9]
```diff
@@ -14,25 +14,16 @@
...
-    const graphClient = clientService.graphAuthenticated(
-      store.getters.configuration.server,
-      store.getters.getToken
-    )
-
-    return { graphClient }
-  },
-  computed: {
-    ...mapGetters(['getToken', 'configuration'])
+    return useGraphClient()
```
Comment (dschmidt) [inline packages/web-app-files/src/components/AppBar/CreateSpace.vue:26]:
```suggestion
    return {
      ...useGraphClient()
    }
```

for consistency's sake...

File: packages/web-runtime/src/services/auth/authService.ts (hunk 4 above)
Comment (fschade) [inline packages/web-runtime/src/services/auth/authService.ts:81]: `console.info`

**Review by fschade — APPROVED** [review]
just a few little things, not critical

File: packages/web-client/src/composables/graphClient/useGraphClient.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+export const useGraphClient = ({ store }: GraphClientOptions = {}) => {
+  store = store || useStore<any>()
+  const accessToken = useAccessToken({ store })
+  const clientService = useClientService()
+
+  const graphClient = computed(() => {
+    return clientService.graphAuthenticated(store.getters.configuration.server, unref(accessToken))
+  })
+
+  return {
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-client/src/composables/graphClient/useGraphClient.ts:18]: Not planned, but I like it as a pattern to always spread the result of composables. Makes it easier to add more later on.

File: packages/web-pkg/src/configuration/manager.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  public initialize(rawConfig: RawConfig): void {
+    this.runtimeConfiguration = { serverUrl: '' }
+    this.serverUrl = rawConfig.server
+    this.oAuth2Configuration = rawConfig.auth ? (rawConfig.auth as OAuth2Configuration) : null
+    this.oidcConfiguration = rawConfig.openIdConnect
+      ? (rawConfig.openIdConnect as OIDCConfiguration)
+      : null
+  }
+
+  set serverUrl(url: string) {
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-pkg/src/configuration/manager.ts:24]: good catch!

File: packages/web-runtime/src/services/auth/authService.ts (hunk 4 above)
Comment (dschmidt) [reply packages/web-runtime/src/services/auth/authService.ts:81]: `UserSignedIn` and `UserSignedOut` are raised from the `SessionMonitor` in `oidc-client-ts`. As we don't use that, we never get this event and should just get rid of it.

**Comment by javfg** [comment]
Hi folks :wave:,

Since you are overhauling the auth flow, I wanted to describe the problem we've been having in a corner case related to auth.

When using oidc-client-js, the token validity check only tests that the expiry date is not met. If we invalidate the token in our SSO (we have a _log out from all applications_ entry in the SSO web), we end up with a broken state in any open CERNBox tabs we have in the browser, as oidc-client-js assumes the token works just because it is not expired. This was taking users to the _missing config file_ page.

We are working around this by cleaning sessionStorage and refreshing the tab if the first authenticated request fails (see [this change](https://github.com/cernbox/web/blob/ec86d3856d5deedc42379b4e383201fe6868cdb7/packages/web-runtime/src/store/user.js#L102:L105)).

Maybe you have a better solution for this with the new library? 

Cheers!

**Comment by kulmann** [comment]
> Hi folks 👋,
> 
> Since you are overhauling the auth flow, I wanted to describe the problem we've been having in a corner case related to auth.
> 
> When using oidc-client-js, the token validity check only tests that the expiry date is not met. If we invalidate the token in our SSO (we have a _log out from all applications_ entry in the SSO web), we end up with a broken state in any open CERNBox tabs we have in the browser, as oidc-client-js assumes the token works just because it is not expired. This was taking users to the _missing config file_ page.
> 
> We are working around this by cleaning sessionStorage and refreshing the tab if the first authenticated request fails (see [this change](https://github.com/cernbox/web/blob/ec86d3856d5deedc42379b4e383201fe6868cdb7/packages/web-runtime/src/store/user.js#L102:L105)).
> 
> Maybe you have a better solution for this with the new library?
> 
> Cheers!

Hey @javfg thanks for reaching out :-) I think that your workaround should not be needed anymore. If an error happens during user info fetching the user gets redirected to a login error page, which explains that the session has become invalid. Could you try it out without your workaround once this PR is merged or even with the current state of the PR? It's in cleanup state and otherwise already finished.

File: packages/web-runtime/src/router/helpers.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+ * @param router {Router}
+ * @param to {Route}
+ * @returns {boolean}
+ */
+export const isUserContext = (router: Router, to: Route): boolean => {
+  if (!isAuthenticationRequired(router, to)) {
+    return false
+  }
+
+  if (to.meta?.auth !== false) {
```
[the remark this answers is not in the document]
Comment (dschmidt) [reply packages/web-runtime/src/router/helpers.ts:47]: @kulmann and I agreed to keep `.auth` and create a follow up ticket.

`__public` is gone now.
